### PR TITLE
Exclude entry points from unused members analysis

### PR DESCRIPTION
**Pull request by sharwell**

### CODE REVIEW

File: src/Analyzers/Core/Analyzers/RemoveUnusedMembers/AbstractRemoveUnusedMembersDiagnosticAnalyzer.cs [hunk 1]
```diff
@@ -689,7 +696,7 @@ private bool IsCandidateSymbol(ISymbol memberSymbol)
             }
 
             private bool IsEntryPoint(IMethodSymbol methodSymbol)
-                => methodSymbol.Name == WellKnownMemberNames.EntryPointMethodName &&
+                => (methodSymbol.Name == WellKnownMemberNames.EntryPointMethodName || methodSymbol.Name == "$Main") &&
```
Comment (sharwell) [inline src/Analyzers/Core/Analyzers/RemoveUnusedMembers/AbstractRemoveUnusedMembersDiagnosticAnalyzer.cs:699]: @AlekseyTs Ideally we would use a helper for "unspeakable names", but I couldn't find one. Can you point us in the right direction?

Comment (AlekseyTs) [reply src/Analyzers/Core/Analyzers/RemoveUnusedMembers/AbstractRemoveUnusedMembersDiagnosticAnalyzer.cs:699]: > Ideally we would use a helper for "unspeakable names", but I couldn't find one. Can you point us in the right direction?

```SyntaxFacts.IsValidIdentifier```?

Comment (AlekseyTs) [reply src/Analyzers/Core/Analyzers/RemoveUnusedMembers/AbstractRemoveUnusedMembersDiagnosticAnalyzer.cs:699]: You also mentioned ```ISymbol.CanBeReferencedByName``` offline. However, it doesn't feel right to include all methods with unspeakable names into entry points.

---
In reply to: [439097752](https://github.com/dotnet/roslyn/pull/45090#discussion_r439097752) [](ancestors = 439097752)

Comment (AlekseyTs) [inline src/Analyzers/Core/Analyzers/RemoveUnusedMembers/AbstractRemoveUnusedMembersDiagnosticAnalyzer.cs:699]: >"$Main" [](start = 107, length = 7)

Perhaps we should move constant from SynthesizedSimpleProgramEntryPointSymbol to WellKnownMemberNames and use it here?

**Comment by AlekseyTs** [comment]
CC @mavasani

File: src/Analyzers/Core/Analyzers/RemoveUnusedMembers/AbstractRemoveUnusedMembersDiagnosticAnalyzer.cs (hunk 1 above)
Comment (sharwell) [reply src/Analyzers/Core/Analyzers/RemoveUnusedMembers/AbstractRemoveUnusedMembersDiagnosticAnalyzer.cs:699]: > Perhaps we should move constant from SynthesizedSimpleProgramEntryPointSymbol to WellKnownMemberNames and use it here?

I'll file a follow-up issue for this.

**Pull request merged**